File: 2022/day2.py
import timeit
file = open('day2_input.txt', 'r')
line = file.read().split("\n")

# Part 1
def part1():
    s = 0
    for x in line:
        if x != '':
            j1, j2 = x.split()
            j1 = ord(j1) - 65
            j2 = ord(j2) - 88
            if j2 == j1:
                s += 3 + j2 + 1
            elif j2 == (j1 + 1) % 3:
                s += j2 + 6 + 1
            else:
                s += j2 + 1
    return s

print(part1())

# x = timeit.default_timer()
# part1()
# print(timeit.default_timer() - x)

# Part 2
def part2():
    s = 0
    for x in line:
        if x != '':
            j1, v = x.split()
            j1 = ord(j1) - 65
            v = ord(v) - 88
            s += (j1 + (v - 1)) % 3 + 1 + v * 3
    return s

print(part2())

# x = timeit.default_timer()
# part2()
# print(timeit.default_timer() - x)

File: cpp/2023/day2/day2.cpp
#include <string>
#include <vector>
#include "fstream"
#include "iostream"
#include "../../splitString.cpp"
#include "unordered_map"

using namespace std;
std::ifstream ifs;



string part1() {
    ifs.open("2023/day2/day2_input.txt");
    string line;
    int score = 0;
    int count = 0;
    getline(ifs,line);
    while (!line.empty()) {

        count ++;
//        cerr << count << line << endl;
        unordered_map <string, int> maxs;
        vector<string> v = splitString(line, ':');
        vector<string> v2 = splitString(v[1], ';');
        for (auto seg : v2) {
            vector<string> v3 = splitString(seg, ',');
            for (auto seg2 : v3) {
                vector<string> v4 = splitString(seg2, ' ');
//                cerr << v4[0] << " " << v4[1] << endl;
                int val = stoi(v4[0]);
                string name = v4[1];
                maxs[name] = max(maxs[name], val);
            }
        }
        if (maxs["red"] <= 12 and maxs["blue"] <= 14 and maxs["green"] <= 13 ) {
            score = score + count;
        }
//        cout << count << ' ' << maxs["red"] << " " << maxs["blue"] << " " << maxs["green"] << endl;
        getline(ifs,line);
    }
//    cout << score << endl;
    ifs.close();
    return to_string(score);
}

string part2() {
    ifs.open("2023/day2/day2_input.txt");
    string line;
    int score = 0;
    int count = 0;
    getline(ifs,line);
    while (!line.empty()) {

        count ++;
//        cerr << count << line << endl;
        unordered_map <string, int> maxs;
        vector<string> v = splitString(line, ':');
        vector<string> v2 = splitString(v[1], ';');
        for (auto seg : v2) {
            vector<string> v3 = splitString(seg, ',');
            for (auto seg2 : v3) {
                vector<string> v4 = splitString(seg2, ' ');
//                cerr << v4[0] << " " << v4[1] << endl;
                int val = stoi(v4[0]);
                string name = v4[1];
                maxs[name] = max(maxs[name], val);
            }
        }

        score = maxs["red"]*maxs["blue"]*maxs["green"] + score;

//        cout << count << ' ' << maxs["red"] << " " << maxs["blue"] << " " << maxs["green"] << endl;
        getline(ifs,line);
    }
//    cout << score << endl;
    ifs.close();
    return to_string(score);
}

#include <benchmark/benchmark.h>

void BM_part1(benchmark::State& state)
{
    while (state.KeepRunning()) {
        string part1_result = part1();
        benchmark::DoNotOptimize(part1_result);
    }
}

void BM_part2(benchmark::State& state)
{
    while (state.KeepRunning()) {
        string part2_result = part2();
        benchmark::DoNotOptimize(part2_result);
    }
}

BENCHMARK(BM_part1);
BENCHMARK(BM_part2);

BENCHMARK_MAIN();

//int main() {
//    cout << part1() << endl;
//    cout << part2() << endl;
//}